awscli/customizations/history/db.py

```diff
@@ -0,0 +1,256 @@
+        # will be routed here instead of the encode method. In the case of
+        # utf-8 bytes we want to make the stored data as readble as possible so
+        # first we try to decode it as utf-8, if that fails we will fall back
+        # to using a base64 encoded string.
+        try:
+            encoded = obj.decode('utf-8')
+        except UnicodeDecodeError:
+            encoded = ensure_text_type(base64.b64encode(obj))
+        return encoded
+
+    def _encode_datetime(self, obj):
+        return obj.isoformat()
+
+    def _unknown(self, obj, type_name):
+        return type_name
+
+    def encode(self, obj):
+        # The default method is not called in PY2 where the JSONEncoder thinks
+        # it can handle a bytes object because it can't tell it apart from a
+        # str object. In PY3 this is handled by the _process_bytes method.
+        # For PY2 we have to override the encode method where the encoding of
+        # a bytes-like string will fail, catch the UnicodeDecodeError and
+        # then call _process_bytes and retry the encoding.
+        try:
+            encoded = super(PayloadSerializer, self).encode(obj)
+            return encoded
+        except UnicodeDecodeError:
+            b64encoded = self._process_bytes(obj)
+            encoded = self.encode(b64encoded)
+            return encoded
+
+    def default(self, obj):
+        if isinstance(obj, binary_type):
+            return self._encode_binary(obj)
+        elif isinstance(obj, datetime.datetime):
+            return self._encode_datetime(obj)
+        elif isinstance(obj, MutableMapping):
+            return self._encode_mutable_mapping(obj)
+        else:
+            return type(obj).__name__
+
+
+class DatabaseRecordWriter(object):
+    _WRITE_RECORD = """
+    INSERT INTO records(id, request_id, source, event_type, timestamp, payload)
+    VALUES (?,?,?,?,?,?)"""
+    _REQUEST_LIFECYCLE_EVENTS = set(
+        ['API_CALL', 'HTTP_REQUEST', 'HTTP_RESPONSE', 'PARSED_RESPONSE'])
+    _START_OF_REQUEST_LIFECYCLE_EVENT = 'API_CALL'
+
+    def __init__(self, connection=None):
+        if connection is None:
+            connection = DatabaseConnection()
+        self._connection = connection
+        self._request_ids = {}
+        self._identifier = None
+
+    def write_record(self, record):
+        # This method is not threadsafe by itself, it is only threadsafe when
+        # used inside a handler bound to the HistoryRecorder in botocore which
+        # is protected by a lock.
+        cursor = self._connection.cursor()
+        db_record = self._create_db_record(record)
+        cursor.execute(self._WRITE_RECORD, db_record)
+        self._connection.commit()
+
+    def _get_request_id(self, event_type):
+        if event_type == self._START_OF_REQUEST_LIFECYCLE_EVENT:
+            self._start_http_lifecycle()
+        if event_type in self._REQUEST_LIFECYCLE_EVENTS:
+            thread_id = threading.current_thread().ident
```

Comment:
tls is the typical way to handle this.  Is there a reason you're not using that?

Reply:
I had it that way initially I changed it due to feedback. I can put it back if you'd prefer. The docs for `ident` say its intended use was for keying a dictionary off of threads so I felt it wasn't too gross this way.